thorlcr/graph/thgraphmaster.cpp

```diff
@@ -390,35 +390,108 @@ CMasterActivity::~CMasterActivity()
 
 IDistributedFile *CMasterActivity::queryReadFile(unsigned f)
 {
-    if (f>=readFiles.ordinality())
-        return NULL;
-    return &readFiles.item(f);
+    if (f>=readFiles.size())
+        return nullptr;
+    return readFiles[f];
 }
 
-IDistributedFile *CMasterActivity::findReadFile(const char *lfnName)
+unsigned CMasterActivity::queryReadFileId(const char *normalizedFileName)
 {
-    auto it = readFilesMap.find(lfnName);
+    auto it = readFilesMap.find(normalizedFileName);
     if (it != readFilesMap.end())
-        return LINK(it->second);
+        return (int)(it->second);
+    return NotFound;
+}
+
+IDistributedFile *CMasterActivity::findReadFile(const char *normalizedFileName)
+{
+    auto it = readFilesMap.find(normalizedFileName);
+    if (it != readFilesMap.end())
+        return LINK(readFiles[it->second]);
     return nullptr;
 }
 
-IDistributedFile *CMasterActivity::lookupReadFile(const char *lfnName, AccessMode mode, bool jobTemp, bool temp, bool opt)
+unsigned CMasterActivity::setupSpace4FileStats(IDistributedFile * file, unsigned fileId, bool statsForMultipleFiles, const StatisticsMapping & statsMapping)
+{
+    unsigned fileStatsStartEntry = 0;
+    unsigned fileStatsSizeExtend = 0; // how many entries need to be added to fileStats vector
+    IDistributedSuperFile *super = file->querySuperFile();
+    bool isSuper = super!=nullptr;
+    unsigned numSubs = isSuper ? super->numSubFiles(true) : 0;
+    if (statsForMultipleFiles)
+    {
+        // fileStats[] will need to track stats from multiple files.  To do this, fileStatsTable[] vector
+        // will be used to map from file id to the starting point of the stats in fileStats.
+        // e.g.
+        // Given that file id for 'child::super' is 0
+        //            file id for 'child::f1' is 1
+        //            file id for 'child::f2' is 2
+        //       and child::super has 2 subfiles
+        //
+        // Then the following table will be generated (through multiple calls to this function)
+        // fileStatsTable[0] = 0;
+        // fileStatsTable[1] = 2;
+        // fileStatsTable[2] = 3;
+        //
+        // fileStats[0] = stats for 'child::super' - stats for child::super and subfile 1
+        // fileStats[1] = stats for 'child::super' - stats for child::super and subfile 2
+        // fileStats[2] = stats for 'child::f1' - stats for child::f1
+        // fileStats[3] = stats for 'child::f2' - stats for child::f2
+        if (fileId >= (unsigned) fileStatsTable.size())
+        {
+            unsigned numberToFill = fileId - fileStatsTable.size() + 1;
+            while (numberToFill--)
+                fileStatsTable.push_back(NotFound); //using NotFound as a marker for unused entry
+        }
+        if (fileStatsTable[fileId] == NotFound)
```

Comment:
as far as I can see, fileId will always be >= than fileStatTable.size(), when this is called from lookupReadFile.
So the condition on 440 and check here on line 446 aren't required.
Code becomes:
```
                if (statsForMultipleFiles)
                {
                    unsigned numberToFill = fileId - fileStatsTable.size() + 1;
                    while (numberToFill--)
                        fileStatsTable.push_back(NotFound); //using NotFound as a marker for unused entry

                    fileStatsStartEntry = (unsigned) fileStats.size(); // index of the first stats entry for this file
                    fileStatsTable[fileId] = fileStatsStartEntry;
                    fileStatsSizeExtend = isSuper?numSubs:1;
                }
```

.. but I think it would be clearer (e.g. that fileId is always new before this is called), if the work setupSpace4FileStats is doing is folded into lookupReadFile.
Might also be tempted to change to use a map for fileStatsTable for clarity - removing need to pad to match readFiles size.